tests/Rule/AbstractRuleValidatorTest.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Tests\Rule;
+
+use PHPUnit\Framework\TestCase;
+use stdClass;
+use Yiisoft\Validator\Exception\UnexpectedRuleException;
+use Yiisoft\Validator\Formatter;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\Rule\RuleHandlerInterface;
+use Yiisoft\Validator\Tests\Stub\FakeValidatorFactory;
+use Yiisoft\Validator\ValidationContext;
+
+abstract class AbstractRuleValidatorTest extends TestCase
+{
+    /**
+     * @dataProvider passedValidationProvider
+     */
+    public function testValidationPassed(object $config, mixed $value): void
+    {
+        $result = $this->validate($value, $config);
+
+        $this->assertTrue($result->isValid(), print_r($result->getErrors(), true));
+    }
+
+    /**
+     * @dataProvider failedValidationProvider
+     */
+    public function testValidationFailed(object $config, mixed $value, array $expectedErrors): void
```

Comment:
Need test value paths also.

Reply:
There test for error object so value paths will be tested as well.

Reply:
We need to test value paths that returns by handlers. Important for `Nested` and `Each`. See https://github.com/yiisoft/validator/blob/45f2ff4029ed970c6ec837e0e8b8959698ad444d/tests/Rule/NestedTest.php#L285